### PR TITLE
Produce a diff (if we can) for roundtrip tests

### PR DESCRIPTION
**Pull request by jystic**
!! @amosr @charleso 

This makes the tripping test in disorder produce a diff (just like `===`) if it is able to parse the string representation of original and roundtrip.

So `tripping` now produces output like this:

``` diff
Roundtrip failed.

=== - Original / + Roundtrip ===
 Right
   Block
   { blockEntities =
     [ BlockEntity
       { entityHash =
         EntityHash 7
       , entityId =
         EntityId "kyle-000"
       , entityAttributes =
-        [ BlockAttribute
-            { attributeId = AttributeId 0 , attributeRows = 0 }
-        ]
+        []
       }
     ]
   , blockIndices =
     []
   , blockTables =
     []
   }
```

If the string representation could not be parsed then it reverts to the old behaviour:

```
Roundtrip failed.

=== Original ===
Right
  Block
    { blockEntities =
        [ BlockEntity
            { entityHash = EntityHash 7
            , entityId = EntityId "kyle-000"
            , entityAttributes =
                [ BlockAttribute
                    { attributeId = AttributeId 0 , attributeRows = 0 }
                ]
            }
        ]
    , blockIndices = []
    , blockTables = []
    }

=== Roundtrip ===
Right
  Block
    { blockEntities =
        [ BlockEntity
            { entityHash = EntityHash 7
            , entityId = EntityId "kyle-000"
            , entityAttributes = []
            }
        ]
    , blockIndices = []
    , blockTables = []
    }
```

### CODE REVIEW

**Comment by jystic** [comment]
!! @erikd-ambiata might be interested too

**Comment by amosr** [comment]
👍 
great

File: disorder-jack/src/Disorder/Jack/Tripping.hs [hunk 1]
```diff
@@ -35,13 +38,25 @@ trippingString render to fro x =
...
     counterexample "" .
     counterexample "Roundtrip failed." .
     counterexample "" .
-    counterexample "=== Original ===" .
-    counterexample (render original) .
-    counterexample "" .
-    counterexample "=== Roundtrip ===" .
-    counterexample (render roundtrip) $
+    counterexample (List.intercalate "\n" $ fromMaybe comparison diff) $
       property (roundtrip == original)
```
Comment (charleso) [inline disorder-jack/src/Disorder/Jack/Tripping.hs]: Dumb question, any reason this isn't just `===`? Would that do the same thing?

**Comment by erikd-ambiata** [comment]
:100: :100: :100:

**Comment by charleso** [comment]
👍  if you can't use `===` directly

File: disorder-jack/src/Disorder/Jack/Tripping.hs (hunk 1 above)
Comment (erikd-ambiata) [inline disorder-jack/src/Disorder/Jack/Tripping.hs]: I do think you're right there!

**Comment by jystic** [comment]
Discussed IRL, `===` required `Show`

**Comment by jystic** [comment]
The other reason is that `===` gives the header `=== Not equal ===` and I wanted to keep the clear output info that tripping has where it tells you which one is the original and which is the roundtrip

**Pull request merged**